# AlleleAnalyzer

AlleleAnalyzer is a tool for designing personalized and allele-specific sgRNAs for general genome editing and optimizing pairs of sgRNAs for human therapeutic editing. This tool was formerly named ExcisionFinder, but ExcisionFinder is now a component of the larger AlleleAnalyzer tool focused on paired sgRNA design.

**Note**:  If you use this tool and find any issues or anything that is difficult to interpret, please send a message to this Github or to kathleen.keough at gladstone.ucsf.edu

If you use this tool, please cite our [preprint](https://www.biorxiv.org/content/early/2018/09/11/342923).

## Table of Contents

1. Getting started
* [How it works](https://github.com/keoughkath/ExcisionFinder/wiki/Overview)
* [Install ExcisionFinder and its requirements](https://github.com/keoughkath/ExcisionFinder/wiki/Install-ExcisionFinder-and-its-requirements)
2. Tool descriptions
* [Design allele-specific sgRNAs](https://github.com/keoughkath/ExcisionFinder/wiki/Usage:-gen_sgRNAs.py)
* [Using CRISPOR with ExcisionFinder](https://github.com/keoughkath/ExcisionFinder/wiki/Using-CRISPOR-with-gen_sgRNAs.py)
3. Tutorials
* [Tutorial: Design sgRNAs for allele specific excision of the gene MFN2 in the WTC genome](https://github.com/keoughkath/ExcisionFinder/wiki/Tutorial:-Design-sgRNAs-for-allele-specific-excision-of-the-gene-MFN2-in-the-WTC-genome)

# AlleleAnalyzer 1000 Genomes analysis dataset

The annotated variant database and gene targetability results from the 1000 Genomes Project are available [here](http://lighthouse.ucsf.edu/public_files_no_password/excisionFinderData_public/ExcisionFinder_manuscript_data/).

## Useful links

[Pre-computed PAM locations.](http://lighthouse.ucsf.edu/public_files_no_password/excisionFinderData_public/)

[UCSC chromosome fastas, hg19.](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/)

[UCSC chromosome fastas, hg38.](http://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/)


